packages/schematics/angular/utility/workspace_spec.ts

```diff
@@ -0,0 +1,110 @@
+import { getWorkspace as readWorkspace, updateWorkspace, writeWorkspace } from './workspace';
+
+const TEST_WORKSPACE_CONTENT = JSON.stringify({
+  version: 1,
+  projects: {
+    'test': {},
+  },
+});
+
+describe('readWorkspace', () => {
+  it('reads a workspace using the default path value', async () => {
+    const tree = new EmptyTree();
+    tree.create('/angular.json', TEST_WORKSPACE_CONTENT);
+
+    const workspace = await readWorkspace(tree);
+    expect(workspace.projects.has('test')).toBeTrue();
+  });
+
+  it('reads a workspace when specifying a directory path', async () => {
+    const tree = new EmptyTree();
+    tree.create('/xyz/angular.json', TEST_WORKSPACE_CONTENT);
+
+    const workspace = await readWorkspace(tree, '/xyz/');
+    expect(workspace.projects.has('test')).toBeTrue();
+  });
+
+  it('reads a workspace when specifying a file path', async () => {
+    const tree = new EmptyTree();
+    tree.create('/angular.json', TEST_WORKSPACE_CONTENT);
+
+    const workspace = await readWorkspace(tree, 'angular.json');
+    expect(workspace.projects.has('test')).toBeTrue();
+  });
+
+  it('throws if workspace file does not exist when using the default path value', async () => {
+    const tree = new EmptyTree();
+
+    await expectAsync(readWorkspace(tree)).toBeRejectedWithError();
+  });
+
+  it('throws if workspace file does not exist when specifying a file path', async () => {
+    const tree = new EmptyTree();
+
+    await expectAsync(readWorkspace(tree, 'angular.json')).toBeRejectedWithError();
+  });
+});
+
+describe('writeWorkspace', () => {
+  it('writes a workspace using the default path value', async () => {
+    const tree = new EmptyTree();
+    tree.create('/angular.json', TEST_WORKSPACE_CONTENT);
+    const workspace = await readWorkspace(tree);
+
+    workspace.extensions['x-abc'] = 1;
+    await writeWorkspace(tree, workspace);
+    expect(tree.readJson('/angular.json')).toEqual(jasmine.objectContaining({ 'x-abc': 1 }));
+  });
+
+  it('writes a workspace when specifying a path', async () => {
+    const tree = new EmptyTree();
+    tree.create('/angular.json', TEST_WORKSPACE_CONTENT);
+    const workspace = await readWorkspace(tree);
+
+    workspace.extensions['x-abc'] = 1;
+    await writeWorkspace(tree, workspace, '/xyz/angular.json');
+    expect(tree.readJson('/xyz/angular.json')).toEqual(jasmine.objectContaining({ 'x-abc': 1 }));
+  });
+});
+
+describe('updateWorkspace', () => {
+  it('updates a workspace using the default path value', async () => {
+    const tree = new EmptyTree();
+    tree.create('/angular.json', TEST_WORKSPACE_CONTENT);
+
+    const rule = updateWorkspace((workspace) => {
+      workspace.projects.add({
+        name: 'abc',
+        root: 'src',
+      });
+    });
+
+    // Simulate execution of the rule in a schematic
+    await rule(tree, {} as unknown as SchematicContext);
+
+    expect(tree.read('angular.json')?.toString()).toContain('"abc"');
```

Comment:
Nit: Could we simplify this assertion (and some of the previous ones) with `readWorkspace()`?